pytorch_lightning/core/optimizer.py

```diff
@@ -184,30 +227,19 @@ def dis_closure():
 
         if make_optimizer_step is None:
             make_optimizer_step = not self._should_accumulate
-
-        trainer = self._trainer
-        optimizer = self._optimizer
+        else:
+            if self._automatic_optimization:
+                raise MisconfigurationException(
+                    'In automatic optimization, `make_optimizer_step` should be None. '
+                    'This option is enabled only with Trainer(automatic_optimization=False) '
+                    'Currently, all optimizers follow the Trainer(accumulate_grad_batches=x) logic')
 
         if make_optimizer_step:
-            if trainer.on_tpu:
-                with trainer.profiler.profile(profiler_name):
-                    xm.optimizer_step(optimizer, optimizer_args={'closure': closure, **kwargs})
-
-            elif trainer.amp_backend is not None:
-                trainer.precision_connector.backend.optimizer_step(
-                    trainer, optimizer, closure)
-
-            else:
-                with trainer.profiler.profile(profiler_name):
-                    optimizer.step(closure=closure, *args, **kwargs)
-
-            # perform zero grad
-            optimizer.zero_grad()
+            self.__optimizer_step(*args, closure=closure, profiler_name=profiler_name, **kwargs)
         else:
             # make sure to call optimizer_closure when accumulating
-            with trainer.profiler.profile("closure"):
-                with trainer.train_loop.block_ddp_sync_behaviour():
-                    closure()
+            with self._trainer.profiler.profile("closure"), self._trainer.train_loop.block_ddp_sync_behaviour():
```

Comment:
shall sec which closure - closure for what optimizer?

Reply:
Great idea !